include/cuco/detail/static_map.inl

```diff
@@ -422,6 +422,83 @@ __device__ bool static_map<Key, Value, Scope, Allocator>::device_mutable_view::i
   }
 }
 
+template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
+template <typename Hash, typename KeyEqual>
+__device__
+  thrust::pair<typename static_map<Key, Value, Scope, Allocator>::device_mutable_view::iterator,
+               bool>
+  static_map<Key, Value, Scope, Allocator>::device_mutable_view::fetch_insert(
+    value_type const& insert_pair, Hash hash, KeyEqual key_equal) noexcept
+{
+  auto current_slot{initial_slot(insert_pair.first, hash)};
+
+  while (true) {
+    key_type const existing_key = current_slot->first.load(cuda::std::memory_order_relaxed);
+    // The user provide `key_equal` can never be used to compare against `empty_key_sentinel` as the
+    // sentinel is not a valid key value. Therefore, first check for the sentinel
+    auto const slot_is_available =
+      detail::bitwise_compare(existing_key, this->get_empty_key_sentinel()) or
+      detail::bitwise_compare(existing_key, this->get_erased_key_sentinel());
+
+    // the key we are trying to insert is already in the map, so we return with failure to insert
+    if (not slot_is_available and key_equal(existing_key, insert_pair.first)) {
+      // If we cannot use a single CAS operation, ensure that the write to
+      // the value part also took place.
+      if constexpr (not cuco::detail::is_packable<value_type>()) {
+        auto& slot_value       = current_slot->second;
+        auto const empty_value = this->get_empty_value_sentinel();
+        while (
+          detail::bitwise_compare(slot_value.load(cuda::std::memory_order_relaxed), empty_value)) {
```

Comment:
I see what you mean now. 

The idea is to prevent another thread from returning an iterator to a slot in a partial state by spinning on the value. 

We'd need to think about this very carefully to make sure there aren't any subtle memory ordering or ABA issues with this approach. I'll make @sleeepyjack do that work for me ;) 

This approach also can't work on anything below Volta as it requires independent thread scheduling, so we'd need to guard it based on the arch (which is fine).

Reply:
> This approach also can't work on anything below Volta as it requires independent thread scheduling, so we'd need to guard it based on the arch (which is fine).

True, the spin loop needs the forwards guarantee that the inserting thread eventually writes the value...

Reply:
Imagine there are 3 threads: T0, T1 and T2 inserting `{5, 1}`, `{5, 2}` and `{5, 3}`
T0 wins key CAS but not value CAS
T1 wins value CAS but not key CAS
So the slot content becomes `{5, 2}`
When T2 comes in, it finds the slot is not available and the value is not sentinel. So it returns a pointer to `{5, 2}` while the expected would be `{5, 1}`.

Did I miss anything?

Reply:
Yes, that's exactly the reason why we cannot use `back_to_back_cas` for implementing `fetch_insert`.

Reply:
Come to think of it, I recall @nicolas-iskos experimented with something like this back when he originally discovered the original race condition that required us to stop returning an iterator. Nico, do you recall why we didn't move forward with that idea?

Reply:
IIRC, I implemented something very similar to this to attempt to support concurrent insert/find. I think at that time, it slowed down both `insert` and `find`. There may have also been another reason, but I can't recall it at the moment. Perhaps one way to measure the perf impact would be to use `packed_cas` for 4B/4B pairs and compare throughput to the `cas_dependent_write` + spin loop?

Reply:
> Perhaps one way to measure the perf impact would be to use packed_cas for 4B/4B pairs and compare throughput to the cas_dependent_write + spin loop?

It would be interesting to see what the perf difference is with this implementation. 

@hahnjo would you be interested in making a benchmark for this new API? It would be nice, but I also don't want to hold this PR back unnecessarily if you'd prefer to just get it merged and we can revisit benchmarking in a future PR.

Reply:
> Perhaps one way to measure the perf impact would be to use packed_cas for 4B/4B pairs and compare throughput to the cas_dependent_write + spin loop?

Note that `fetch_insert` will still use `packed_cas` if it can, and in that case the performance should be identical to `insert`.

> @hahnjo would you be interested in making a benchmark for this new API?

I tried to do some benchmarking last week, but IIRC they all use the host APIs which we don't have yet for this new function. So I temporarily tried replacing the functionality of the `insert` host API with the new `fetch_insert`, and I got very weird results: For a start, `cas_dependent_write` seems to be faster than `back_to_back_cas` on my RTX 2070 SUPER. Then having the additional spin loops for `fetch_insert` wasn't much slower.

> It would be nice, but I also don't want to hold this PR back unnecessarily if you'd prefer to just get it merged and we can revisit benchmarking in a future PR.

Yes, I'd prefer to get this merged as-is. If somebody has a good idea how to benchmark this, I'm happy to provide my thoughts on a follow-up PR.

Reply:
>  I got very weird results: For a start, cas_dependent_write seems to be faster than back_to_back_cas on my RTX 2070 SUPER

We've never robustly tested on which platforms `back_to_back_cas` is faster than `cas_dependent_write`. I think we've only really checked V100 and A100, so it's very possible that the latter could be faster on some architectures or with GDDR instead of HBM. 

> Yes, I'd prefer to get this merged as-is. 

Sounds good.